README.md
# Mental Health Service Finder
> Helping to connect people to mental health services.

[![GitHub Release](https://img.shields.io/github/release/nhsuk/mental-health-service-finder.svg)](https://github.com/nhsuk/mental-health-service-finder/releases/latest/)
[![Greenkeeper badge](https://badges.greenkeeper.io/nhsuk/mental-health-service-finder.svg)](https://greenkeeper.io/)
[![Build Status](https://travis-ci.org/nhsuk/mental-health-service-finder.svg?branch=master)](https://travis-ci.org/nhsuk/mental-health-service-finder)
[![Coverage Status](https://coveralls.io/repos/github/nhsuk/mental-health-service-finder/badge.svg?branch=master)](https://coveralls.io/github/nhsuk/mental-health-service-finder?branch=master)

## Installation

Clone the repo: `git clone https://github.com/nhsuk/mental-health-service-finder.git`
and review the [`scripts`](scripts) to get up and running.

## Testing

The application uses [Docker](https://www.docker.com/) to run in containers.
Development is typically done on the host machine. Files are loaded into the
container and changes are automatically updated.

Use the `test` script for continuous testing during development.

## Test environments

As the application is being developed, every Pull Request has its own test
environment automatically built and deployed to.

Every environment apart from the one we want the public to access requires
basic authentication to access. The username and password are not secret, in
fact they are included within environment variable table below.
The intention with the authentication challenge is to prevent people whom may
stumble across the site and not realise it is for testing, it also prevents
access by search engines and other bots.

## Environment variables

Environment variables are expected to be managed by the environment in which
the application is being run. This is best practice as described by
[twelve-factor](https://12factor.net/config).

In order to protect the application from starting up without the required
env vars in place
[require-environment-variables](https://www.npmjs.com/package/require-environment-variables)
is used to check for the env vars that are required for the application to run
successfully.
This happens during the application start-up. If an env var is not found the
application will fail to start and an appropriate message will be displayed.

Environment variables are used to set application level settings for each
environment.

| Variable                           | Description                                                                                         | Default                                             | Required   |
| :--------------------------------- | :-------------------------------------------------------------------------------------------------- | :---------------------------------------------------| :--------- |
| `ADOBE_TRACKING_URL`               | [Adobe Analytics](https://www.adobe.com/analytics/adobe-analytics.html) Dynamic Tag Management URL  |                                                     | No         |
| `BASIC_AUTH`                       | An MD5 encrypted [htpasswd](https://httpd.apache.org/docs/2.4/misc/password_encryptions.html)       | See value set in Vault for `BASIC_AUTH` in defaults |            |
| `COOKIEBOT_SCRIPT_URL`             | The URL for the in-house implementation of Cookiebot                                                | //assets.nhs.uk/scripts/cookie-consent.js           |            |
| `HOTJAR_ANALYTICS_TRACKING_ID`     | [Hotjar](https://www.hotjar.com/) tracking id                                                       |                                                     | No         |
| `LOG_LEVEL`                        | Numeric [log level](https://github.com/trentm/node-bunyan#levels)                                   | Depends on `NODE_ENV`                               |            |
| `NODE_ENV`                         | Node environment                                                                                    | development                                         |            |
| `PORT`                             | Server port                                                                                         | 3000                                                |            |
| `SEARCH_API_HOST`                  | Host name for the [NHS Developer API](https://developer.api.nhs.uk/)                                | api.nhs.uk                                          | Yes        |
| `SEARCH_API_KEY`                   | `subscription-key` for the [NHS Developer API](https://developer.api.nhs.uk/)                       |                                                     | Yes        |
| `SEARCH_API_VERSION`               | Version of the [NHS Developer API](https://developer.api.nhs.uk/)                                   | 1                                                   | Yes        |
| `HEADER_FOOTER_API_ENDPOINT`       | JSON endpoint for NHS.UK header and footer API                                                      | `https://nhsuk-apim-dev-uks.azure-api.net/menu/?api-version=1` | Yes        |
| `HEADER_FOOTER_API_CACHE_TIMEOUT`  | Time in seconds for the expiry of the header and footer data cache                                  | 300                                                 | Yes        |
| `HEADER_FOOTER_API_KEY`            | `subscription-key` needed for higher environments of the NHS.UK header and footer API               |                                                     | No         |

## FAQ

* Is the application failing to start?
  * Ensure all expected environment variables are available within the environment
  * If set, `LOG_LEVEL` must be a number and one of the defined [log levels](https://github.com/trentm/node-bunyan#levels)
  * Check for messages in the logs

## Architecture Decision Records

This repo uses
[Architecture Decision Records](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions)
to record architectural decisions for this project.
They are stored in [doc/adr](doc/adr).


CHANGELOG.md
0.30.0 / 2020-10-07
===================
- Update NHSUK Frontend

0.29.0 / 2020-12-17
===================
-- Temporary note to advise users in London to pick service for their borough.

0.28.0 / 2020-12-08
===================
- Add header and footer variables to Rancher

0.27.0 / 2020-12-01
===================
- Update NHSUK Frontend
- Build header and footer from APIs

0.26.0 / 2020-08-26
===================
- Page title updates

0.25.0 / 2020-08-11
===================
- Start page content updates

0.24.0 / 2020-02-14
===================
- Update npm dependencies
- Update basic auth creds
- Remove new service notification banner

0.23.0 / 2019-11-18
===================
- Use `nhsuk-frontend` for views, scss and JS
- Remove redundant JS and SCSS replaced by `nhsuk-frontend`
- Update site root to include `/service-search`

0.22.0 / 2019-10-31
===================
- Remove Greenkeeper - it is done automatically by the service
- Remove special case for West Hampshire CCG
- Remove special case for Redbridge CCG
- Remove special case for Tower Hamlets CCG
- Remove special case for North Cumbria CCG

0.21.0 / 2019-10-11
===================
- Replace (deprecated) `istanbul` with `nyc`
- Use service specific contact details

0.20.0 / 2019-09-26
===================
- Add footer link to beta cookie policy page 

0.19.0 / 2019-09-20
===================
- Update npm dependencies
- Upgrade Docker container to `node:10.16.3-alpine`
- Add West Hampshire CCG IAPT service

0.18.0 / 2019-07-18
===================
- Update npm dependencies
- Use full path for Open Graph image

0.17.0 / 2019-07-18
===================
- Update npm dependencies
- Upgrade Docker container to use latest node LTS
- Add additional (and update existing) metadata properties to meet minimum set
- Allow Adobe and Hotjar to connect to additional domains

0.16.0 / 2019-03-20
===================
- Upgrade Docker container to `node:10.15.3-alpine`
- Update npm dependencies
- Remove simple cookie banner
- Remove Webtrends (no longer used)
- Remove GA (no longer used)

0.15.0 / 2019-01-11
===================
- Update npm dependencies
- Enable debugging
- Upgrade Docker container to `node:10.15.0-alpine`
- Access APIs via APIM instead of going direct
- Use async/await pattern for API requests

0.14.0 / 2018-11-29
===================
- Upgrade Docker container to `node:10.14.0-alpine`
- Update npm dependencies
- Update Redbridge CCG details

0.13.0 / 2018-11-20
===================
- Update npm dependencies
- Create `digitalData` object for use by Adobe Analytics
- Update CSP for compatibility with Adobe Analytics

0.12.0 / 2018-10-09
===================
- Prevent IE 11 from focusing on SVGs
- Update npm dependencies

0.11.0 / 2018-10-03
===================
- Update npm dependencies
- Multiple accessibility improvements
- Upgrade to `node:8.12.0-alpine`
- Fix skip to content for voice over
- Add Tower Hamlets CCG IAPT service

0.10.0 / 2018-09-14
===================
- Add Adobe Analytics
- Update npm dependencies

0.9.0 / 2018-09-12
==================
- Be specific about service being England only

0.8.0 / 2018-09-07
==================
- Do not link to beta home page
- Track GP search link
- Update npm dependencies
- Update 'contact-us' footer link
- Update 'care and support' header link

0.7.0 / 2018-08-30
==================
- Update cookie policy link
- Remove Heatmaps from Webtrends (as the feature has been discontinued)
- Update python used in container
- Anonymise IP used by Webtrends
- Update favicon for latest design
- Match breadcrumb text with URL slug
- Don't display full website URL
- Don't use abbreviation for telephone
- Update footer links to new pages

0.6.0 / 2018-08-22
==================
- Fix typo
- Cleanup typeahead results display
- Update content for improved usability
- Remove Sign Health from results (a fallback should the data not be correct)
- Use local data for North Cumbria and Redbridge CCGs

0.5.0 / 2018-08-20
==================
- Use latest footer font size
- Update print css for new design
- Update IE css for new design
- Update npm dependencies
- Add urgent help call-out to 'check' page
- Do not display online referral message when no option available
- Hyperlink search again link on no results page
- Add extra info to search page about use of GP
- Update content to latest tested
- Give each page a unique title
- Set background to blue
- Replace references to `NHS.UK` in favour of `NHS`
- Update header search placeholder text
- Remove aria-labelledby on the aria-hidden NHS logo
- Do not highlight via Azure Search Service (because the full word is
  highlighted rather than just the entered term)
- Highlight with hand-rolled solution
- Upgrade to `node:8.11.4-alpine`
- Remove hardcoded api-key
- Fix 1 GP result bug
- Search `Address1` and `Address2` fields in typeahead
- Improve display of typeahead results
- Order IAPT results by distance from GP
- Use specific page name for content page WebTrends tracking
- Add login link to footer
- Add clickable menu item icon

0.4.0 / 2018-07-30
==================
- Typeahead enhancements
- Performance tests
- Use new API

0.3.0 / 2018-07-27
==================
- Restrict scope of typeaheads
- Maintain search query for `search again` link back
- Go to IAPT result page when GP has been selected from typeahead
- Remove `no suggestions` display for zero typeahead results
- Show 5 results in typeahead
- Do not search postcode in typeahead
- Scroll input to top for smaller screen sizes
- Add Meta tags for WebTrends
- Turn off auto 'things' on input
- Use new API
- Add Tracking for WebTrends

0.2.0 / 2018-07-23
==================
- Add typeahead for GP search
- Encode queries for back links
- Improve accessibility
- Content and design tweaks
- Increase test coverage
- Extend CSP for Hotjar
- Do not use `role="button"` for links
- Break long terms to prevent overflow
- Improve page heading hierarchy
- Improve aria labelling
- Use error message in page title

0.1.1 / 2018-07-10

config/config.js
const path = require('path');

const rootPath = path.normalize(`${__dirname}/..`);

module.exports = {
  analytics: {
    adobeTrackingUrl: process.env.ADOBE_TRACKING_URL,
    hotjarId: process.env.HOTJAR_ANALYTICS_TRACKING_ID,
  },
  app: {
    env: process.env.NODE_ENV || 'development',
    name: 'mental-health-service-finder',
    port: process.env.PORT || 3000,
    root: rootPath,
  },
  cookiebot: {
    scriptUrl: process.env.COOKIEBOT_SCRIPT_URL || '//assets.nhs.uk/scripts/cookie-consent.js',
  },
  headerFooterApi: {
    cacheTimeout: process.env.HEADER_FOOTER_API_CACHE_TIMEOUT || 300,
    endpoint: process.env.HEADER_FOOTER_API_ENDPOINT || 'https://nhsuk-apim-dev-uks.azure-api.net/menu/?api-version=1',
    subscriptionKey: process.env.HEADER_FOOTER_API_KEY || null,
  },
  search: {
    apiKey: process.env.SEARCH_API_KEY,
    host: process.env.SEARCH_API_HOST || 'api.nhs.uk',
    suggesters: {
      organisation: 'orgname-suggester',
    },
    version: process.env.SEARCH_API_VERSION || '1',
  },
};
